diff --git a/src/main/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java b/src/main/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java
index ee2abb26cf..69607104aa 100644
--- a/src/main/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java
+++ b/src/main/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java
@@ -1608,7 +1608,7 @@ public class StaticTypeCheckingVisitor extends ClassCodeVisitorSupport {
             }
         } else {
             methods = receiver.getMethods(name);
-            if (methods.isEmpty() && args==null || args.length==0) {
+            if (methods.isEmpty() && (args==null || args.length==0)) {
                 // check if it's a property
                 String pname = null;
                 if (name.startsWith("get")) {
@@ -1625,6 +1625,21 @@ public class StaticTypeCheckingVisitor extends ClassCodeVisitorSupport {
 
                     }
                 }
+            } else if (methods.isEmpty() && args!=null && args.length==1) {
+                // maybe we are looking for a setter ?
+                if (name.startsWith("set")) {
+                    String pname = java.beans.Introspector.decapitalize(name.substring(3));
+                    PropertyNode property = receiver.getProperty(pname);
+                    if (property != null) {
+                        ClassNode type = property.getOriginType();
+                        if (implementsInterfaceOrIsSubclassOf(args[0], type)) {
+                            return Collections.singletonList(
+                                    new MethodNode(name, Opcodes.ACC_PUBLIC, VOID_TYPE, new Parameter[]{
+                                            new Parameter(type, "arg")
+                                    }, ClassNode.EMPTY_ARRAY, EmptyStatement.INSTANCE));
+                        }
+                    }
+                }
             }
         }
 
diff --git a/src/test/groovy/transform/stc/FieldsAndPropertiesSTCTest.groovy b/src/test/groovy/transform/stc/FieldsAndPropertiesSTCTest.groovy
index dde04467ee..3584a2af0e 100644
--- a/src/test/groovy/transform/stc/FieldsAndPropertiesSTCTest.groovy
+++ b/src/test/groovy/transform/stc/FieldsAndPropertiesSTCTest.groovy
@@ -193,6 +193,24 @@ class FieldsAndPropertiesSTCTest extends StaticTypeCheckingTestCase {
         '''
     }
 
+    // GROOVY-5232
+    void testSetterForProperty() {
+        assertScript '''
+            class Person {
+                String name
+
+                static Person create() {
+                    def p = new Person()
+                    p.setName("Guillaume")
+                    // but p.name = "Guillaume" works
+                    return p
+                }
+            }
+
+            Person.create()
+        '''
+    }
+
     public static class BaseClass {
         int x
     }
