diff --git a/src/main/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java b/src/main/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java
index a3b5e95c65..a47eedc610 100644
--- a/src/main/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java
+++ b/src/main/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java
@@ -1874,7 +1874,7 @@ public class StaticTypeCheckingVisitor extends ClassCodeVisitorSupport {
         }
 
         // Divisions may produce different results depending on operand types
-        if (DIVIDE == op || DIVIDE_EQUAL == op) {
+        if (isNumberCategory(getWrapper(leftRedirect)) && (DIVIDE == op || DIVIDE_EQUAL == op)) {
             if (isFloatingCategory(leftRedirect) || isFloatingCategory(rightRedirect)) {
                 if (!isPrimitiveType(leftRedirect) || !isPrimitiveType(rightRedirect)) {
                     return Double_TYPE;
diff --git a/src/test/groovy/transform/stc/BugsSTCTest.groovy b/src/test/groovy/transform/stc/BugsSTCTest.groovy
new file mode 100644
index 0000000000..a0f3ec9410
--- /dev/null
+++ b/src/test/groovy/transform/stc/BugsSTCTest.groovy
@@ -0,0 +1,16 @@
+package groovy.transform.stc
+
+/**
+ * Unit tests for static type checking : bug fixes.
+ *
+ * @author Cedric Champeau
+ */
+class BugsSTCTest extends StaticTypeCheckingTestCase {
+    // GROOVY-5456
+    void testShouldNotAllowDivOnUntypedVariable() {
+        shouldFailWithMessages '''
+            def foo(Closure cls) {}
+            def bar() { foo { it / 2 } }
+        ''', 'Cannot find matching method java.lang.Object#div(int)'
+    }
+}
diff --git a/src/test/org/codehaus/groovy/classgen/asm/sc/BugsStaticCompileTest.groovy b/src/test/org/codehaus/groovy/classgen/asm/sc/BugsStaticCompileTest.groovy
new file mode 100644
index 0000000000..f9a87eace5
--- /dev/null
+++ b/src/test/org/codehaus/groovy/classgen/asm/sc/BugsStaticCompileTest.groovy
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2003-2010 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.codehaus.groovy.classgen.asm.sc
+
+import groovy.transform.stc.BugsSTCTest
+
+/**
+ * Unit tests for static type checking : bugs.
+ *
+ * @author Cedric Champeau
+ */
+@Mixin(StaticCompilationTestSupport)
+class BugsStaticCompileTest extends BugsSTCTest {
+
+    @Override
+    protected void setUp() {
+        super.setUp()
+        extraSetup()
+    }
+}
+
